### DISCUSSION TITLE
How to add Python support in a Theia application

### DISCUSSION THREAD

**Discussion by fipro78** [eclipse-theia/theia#11498] [Q&A]
Hi,

I am trying to create a Theia application that contains Python support. Unfortunately I don't get it to work. 

### TLDR

As a short summary of the issues I faced:
- @theia/python extension seems to be outdated and doesn't work anymore
- Newer VS Code Python extensions require Pylance which is not available in OpenVSX
- Versions => 2021 of the VS Code Python extension require the VS Code Engine >= 1.57.0, which is not yet supported in Theia
- Proxy settings seem to not work for installing VS Code Extensions
- Python not recognized

Now the further details for whom is interested. I have tried the following:

### Use [@theia/python](https://www.npmjs.com/package/@theia/python)

I tried this because it is mentioned in [Build your own IDE/Tool](https://theia-ide.org/docs/composing_applications/).
With a current Theia version this results in several strange looking compilation errors at build time. I suppose this is because the extension is several years old and was not updated anymore. So probably it was created with an older Python version, and in a current container with a newer Python version it breaks. At least that is my assumption, not sure if that is correct.

Anyhow, if the @theia/python extension is not further developed and does not work with newer Theia versions, it should probably not be mentioned in the documentation. It took me quite a while to get to the point that the extension seems to be outdated.

### Use the current VS Code Extension as built-in

As Theia supports the usage of VS Code Extensions, I removed the @theia/python extension and added the ms-python VS Code extension from open-vsx.org. I added it as built-in extension and tried to use the newest version available. This fails for several reasons:

- it tries to resolve dependencies from the VS Code Marketplace (ms-toolsai.jupyter,ms-python.vscode-pylance), which of course fails. I can add the extensions to theiaPluginsExcludeIds, but then the necessary extensions are missing. I could add some of them explicitly from open-vsx.org, but it seems that pylance is not available on open-vsx. I don't know if it is possible to add pylance to open-vsx aswell, but it looks like the current Python extension in openvsx is useless for Theia as dependencies are not available.
- excluding the extensions resulted in a succeeding build, but I get several issues at runtime. Most of the errors are about missing submenus (e.g. _Error: Could not find submenu with id editor/title/run_) and there is this error _Unsupported activation events: onWalkthrough:pythonWelcome, onWalkthrough:pythonDataScienceWelcome_

### Use the Python VS Code extension installed at runtime

My guess is that the newest Python extension available requires a newer VS Code Engine. To find a matching version I therefore removed the built-in plugin and tried to install the extension in the resulting Theia application at runtime. The Extensions tab offered me then the version 2020.10.332292344, which by the way seems to be the last version without the pylance dependency. After some investigation I noticed that the next version 2021.8.1159798656 already requires the VS Code Engine 1.57.0, which raised the question, which engine is supported in Theia? I only found the answer here: https://eclipse-theia.github.io/theia/docs/next/modules/plugin_ext_vscode.html

Maybe it would be a good idea to have that information also somewhere in the documentation. Maybe the README of plugin-ext-vscode?

Anyhow, the installation at runtime fails with this exception:

```
2022-08-03T09:09:51.692Z root INFO [ms-python.python]: trying to resolve latest version...
2022-08-03T09:09:52.687Z root INFO [ms-python.python]: resolved to 'ms-python.python-2020.10.332292344'
2022-08-03T09:09:52.688Z root INFO [ms-python.python-2020.10.332292344]: trying to download from "https://open-vsx.org/api/ms-python/python/2020.10.332292344/file/ms-python.python-2020.10.332292344.vsix"... to path /home/node/.theia/extensions
2022-08-03T09:09:59.226Z root ERROR Failed to resolve plugins from 'vscode:extension/ms-python.python' Error: ENOENT: no such file or directory, open '/home/node/.theia/extensions/ms-python.python-2020.10.332292344.vsix'
2022-08-03T09:09:59.228Z root INFO Deploy plugin vscode:extension/ms-python.python: 7541.3 ms [Finished 74.010 s after backend start]
2022-08-03T09:09:59.228Z root ERROR Request deploy failed with error: Deployment of extension with ID vscode:extension/ms-python.python failed. Error: Deployment of extension with ID vscode:extension/ms-python.python failed.
    at PluginServerHandler.deploy (/workspaces/rds_visualization_app/node_modules/@theia/plugin-ext/lib/main/node/plugin-server-handler.js:41:19)
    at async JsonRpcProxyFactory.onRequest (/workspaces/rds_visualization_app/node_modules/@theia/core/lib/common/messaging/proxy-factory.js:132:24)
    at async RpcProtocol.handleRequest (/workspaces/rds_visualization_app/node_modules/@theia/core/lib/common/message-rpc/rpc-protocol.js:157:28)
2022-08-03T09:09:59.265Z root INFO [70b0794e-1b22-4050-a36d-f9e67c83d644] Waiting for backend deployment: 28.1 ms [Finished 36.199 s after frontend start]
2022-08-03T09:09:59.267Z root INFO [70b0794e-1b22-4050-a36d-f9e67c83d644] Sync of 0 plugins: 0.0 ms [Finished 36.199 s after frontend start]
2022-08-03T09:09:59.267Z root INFO [70b0794e-1b22-4050-a36d-f9e67c83d644] Load contributions of 0 plugins: 0.1 ms [Finished 36.199 s after frontend start]
2022-08-03T09:09:59.293Z root INFO [70b0794e-1b22-4050-a36d-f9e67c83d644] Start of 0 plugins: 21.8 ms [Finished 36.230 s after frontend start]
```

I am behind a corporate proxy. As I can access the OpenVSX registry for showing available extensions and even resolving the latest version works, the internet connection via proxy works in general. But it looks like the extension installation then fails with some proxy issue. Is this a known issue? Is it possible to work around this with some additional setting maybe?

### Use the matching Python VS Code extension as built-in

As I now know the version that can be integrated, I tried to include the Python Extension as built-in. The build succeeds. The Theia application starts and can be opened in a browser. It shows the following error on startup:

```
2022-08-03T09:22:49.843Z root WARN Failed to register a menu item for plugin ms-python.python contributed to debug/variables/context {
  command: 'python.datascience.showDataViewer',
  submenu: undefined,
  alt: undefined,
  group: '1_view',
  when: "python.isDebuggerDataViewerExperimentEnabled && debugProtocolVariableMenuContext == 'viewableInDataViewer'"
} Error: Could not find submenu with id debug/variables/context
```

But that is probably not critical.

And I see this log entry:

`Unsupported activation events: onNotebook:jupyter-notebook, please open an issue: https://github.com/eclipse-theia/theia/issues/new`

The problem is that although the Python Extension seems to be installed correctly, it doesn't really work. Python is not recognized. And if I try to set the Python interpreter via _Python: Select Interpreter_ I get the following error:

```
2022-08-03T09:31:59.487Z root ERROR [hosted-plugin: 1827] Promise rejection not handled in one second: Error: Cannot set configuration for unknown language python , reason: Error: Cannot set configuration for unknown language python
With stack trace: Error: Cannot set configuration for unknown language python
    at Object.setLanguageConfiguration (http://localhost:3000/vendors-node_modules_theia_core_shared_phosphor_domutils_index_js-node_modules_theia_monaco-e-1a4eba.js:154858:15)
    at LanguagesMainImpl.$setLanguageConfiguration (http://localhost:3000/vendors-node_modules_theia_plugin-ext_lib_hosted_browser_hosted-plugin_js.js:5295:48)
    at RPCProtocolImpl.doInvokeHandler (http://localhost:3000/vendors-node_modules_theia_plugin-ext_lib_common_disposable-util_js-node_modules_theia_plugin-cba7ee.js:1394:23)
    at RPCProtocolImpl.invokeHandler (http://localhost:3000/vendors-node_modules_theia_plugin-ext_lib_common_disposable-util_js-node_modules_theia_plugin-cba7ee.js:1379:41)
    at RPCProtocolImpl.receiveRequest (http://localhost:3000/vendors-node_modules_theia_plugin-ext_lib_common_disposable-util_js-node_modules_theia_plugin-cba7ee.js:1343:33)
    at RPCProtocolImpl.receiveOneMessage (http://localhost:3000/vendors-node_modules_theia_plugin-ext_lib_common_disposable-util_js-node_modules_theia_plugin-cba7ee.js:1307:26)
    at http://localhost:3000/vendors-node_modules_theia_plugin-ext_lib_common_disposable-util_js-node_modules_theia_plugin-cba7ee.js:1231:48
    at http://localhost:3000/bundle.js:58560:69
    at CallbackList.invoke (http://localhost:3000/bundle.js:58566:26)
    at Emitter.fire (http://localhost:3000/bundle.js:58681:29)
```

Does anybody have a working example on how to build a current Theia application with Python support?
Does anybody have an idea on what I am doing wrong and why the Python support is not working in my example?

### Additional Information

Operating System: Windows
Development Environment: VS Code with Remote Container
Theia Version: 1.28.0
Proxy: Corporate Proxy, locally [px](https://github.com/genotrance/px) installed to avoid authentication in every tool

The following is the Dockerfile for the Remote Container, in which the Theia application then also starts:

```
# See here for image contents: https://github.com/microsoft/vscode-dev-containers/tree/v0.231.5/containers/typescript-node/.devcontainer/base.Dockerfile

# Node.js version (use -bullseye variants on local arm64/Apple Silicon): 16, 14, 12, 16-bullseye, 14-bullseye, 12-bullseye, 16-buster, 14-buster, 12-buster
ARG VARIANT="14"
FROM mcr.microsoft.com/vscode/devcontainers/typescript-node:0-${VARIANT}

ENV http_proxy http://host.docker.internal:3128
ENV https_proxy http://host.docker.internal:3128
ENV GLOBAL_AGENT_HTTP_PROXY http://host.docker.internal:3128
ENV GLOBAL_AGENT_HTTPS_PROXY http://host.docker.internal:3128
ENV ELECTRON_GET_USE_PROXY true

# Theia dependencies
RUN apt-get update && export DEBIAN_FRONTEND=noninteractive \
     && apt-get -y install --no-install-recommends \
     make \
     gcc \
     pkg-config \
     build-essential \
     python3.10 \
     python3-pip \
     software-properties-common \
     libx11-dev \
     libxkbfile-dev \
     libsecret-1-dev \
     libssl-dev

RUN python -m pip install --upgrade pip
RUN pip install torch torchvision torchaudio --extra-index-url https://download.pytorch.org/whl/cpu
RUN pip install torch-scatter torch-sparse torch-cluster torch-spline-conv torch-geometric -f https://data.pyg.org/whl/torch-1.12.0+cpu.html
RUN pip install numpy random2 neomodel
RUN pip install 'python-language-server[all]'

# install code generator for vs code, theia extensions, theia plugins
RUN npm install -g yo code generator-theia-extension @theia/generator-plugin

USER node

EXPOSE 3000
```

The following is the package.json of the browser-app:

```
{
  "private": true,
  "name": "browser-app",
  "version": "0.0.0",
  "dependencies": {
    "@theia/core": "1.28.0",
    "@theia/editor": "1.28.0",
    "@theia/filesystem": "1.28.0",
    "@theia/git": "1.28.0",
    "@theia/markers": "1.28.0",
    "@theia/messages": "1.28.0",
    "@theia/mini-browser": "1.28.0",
    "@theia/monaco": "1.28.0",
    "@theia/navigator": "1.28.0",
    "@theia/outline-view": "1.28.0",
    "@theia/plugin-ext": "1.28.0",
    "@theia/plugin-ext-vscode": "1.28.0",
    "@theia/preferences": "1.28.0",
    "@theia/preview": "1.28.0",
    "@theia/process": "1.28.0",
    "@theia/terminal": "1.28.0",
    "@theia/search-in-workspace": "1.28.0",
    "@theia/vsx-registry": "1.28.0",
    "@theia/workspace": "1.28.0"
  },
  "devDependencies": {
    "@theia/cli": "1.28.0"
  },
  "scripts": {
    "prepare": "yarn run clean && yarn build && yarn run download:plugins",
    "clean": "theia clean",
    "build": "theia build --mode development",
    "start": "theia start --plugins=local-dir:plugins",
    "watch": "theia build --watch --mode development",
    "download:plugins": "theia download:plugins"
  },
  "theiaPluginsDir": "plugins",
  "theiaPlugins": {
    "vscode-python": "https://open-vsx.org/api/ms-python/python/2020.10.332292344/file/ms-python.python-2020.10.332292344.vsix"
  },
  "theiaPluginsExcludeIds": [
  ],
  "theia": {
    "target": "browser"
  }
}
```

**Comment by msujew** [comment]
Hey @fipro78,

> @theia/python extension seems to be outdated and doesn't work anymore

Yes, that's by design, since it predates the vscode extension support (same with the other `@theia/*` language packages). Given that you can now reuse vscode extensions to provide language support for Theia, they are no longer supported.

> Newer VS Code Python extensions require Pylance which is not available in OpenVSX

Yep, Microsoft [made the Pylance extension proprietary](https://marketplace.visualstudio.com/items?itemName=ms-python.vscode-pylance), preventing us from uploading it to open-vsx.

> Proxy settings seem to not work for installing VS Code Extensions

You mean when running `theia download:plugins` or during the runtime of the app? As of now Theia doesn't have access to the necessary data to automatically determine the OS proxy, which is why you need to set the `--proxy-url` argument or `http.proxy` preference respectively for any proxy settings to take effect.

> Python not recognized

I cannot reproduce this on Theia `1.28.0` and python extension version `v2020.10.332292344`. The extension seems to work as expected for me:

![image](https://user-images.githubusercontent.com/4377073/182601764-0f8ec796-e9a1-41cf-8817-1311f19a05f8.png)

**Comment by fipro78** [comment]
> You mean when running theia download:plugins or during the runtime of the app?

At runtime. I have added the details and the logged error in "Use the Python VS Code extension installed at runtime" 

The preferences say "If not set, will be inherited from the http_proxy and https_proxy environment variables.". These environment variables are set, and I guess they are working as OpenVSX can be contacted. It is just not possible to download and install. 

> I cannot reproduce this on Theia 1.28.0 and python extension version v2020.10.332292344. The extension seems to work as expected for me

Hm, have you installed the Python Language Server or is this done automatically? Maybe I have an issue in that area with my dev container.

**Reply by msujew** [reply]
> The preferences say "If not set, will be inherited from the http_proxy and https_proxy environment variables.". These environment variables are set, and I guess they are working as OpenVSX can be contacted. It is just not possible to download and install.

Are you running Theia in the browser or as an Electron app? In the browser case, requests to `OpenVSX` are not tunneled through Theia's own proxy service but instead use the browser's native proxy capabilities. Downloading is done in the backend, where the proxy variables are needed. However, setting these env variables should work. See 

https://github.com/eclipse-theia/theia/blob/7e177f92c6d3f32dbc876d25ec380bd403e13858/dev-packages/request/src/proxy.ts#L23-L31

Are you able to debug Theia and see what's happening in [this request function](https://github.com/eclipse-theia/theia/blob/7e177f92c6d3f32dbc876d25ec380bd403e13858/dev-packages/request/src/node-request-service.ts#L95-L104)? The options/agent object should pick up the env variables if everything's running as expected. 

> Hm, have you installed the Python Language Server or is this done automatically?

I have done nothing except for installing it using the vsx-registry widget.

**Reply by fipro78** [reply]
> Are you running Theia in the browser or as an Electron app?

In the browser

> Are you able to debug Theia and see what's happening

Not sure how to debug. I'm still trying to get my head around Theia itself and try to create a usable Theia application. If you have a pointer how to start debugging, please let me know.

I finally noticed the real issue. It wasn't a proxy issue. I was mislead by this part in the error "JsonRpcProxyFactory.onRequest". The error was that the folder _~/.theia/extensions_ couldn't be created. After I created it manually from a terminal the installation worked.

Maybe it would be good if the error would be more verbose in this method:
https://github.com/eclipse-theia/theia/blob/7e177f92c6d3f32dbc876d25ec380bd403e13858/packages/plugin-ext/src/main/node/plugin-server-handler.ts#L33-L43

One problem solved. But I still have the issue that I can't add additional language support. I now also tried to install Typescript and Javascript support. But I always get the following error (example for python):

```
2022-08-03T13:26:22.975Z root ERROR [hosted-plugin: 15175] Promise rejection not handled in one second: Error: Cannot set configuration for unknown language python , reason: Error: Cannot set configuration for unknown language python
With stack trace: Error: Cannot set configuration for unknown language python
    at Object.setLanguageConfiguration (http://localhost:3000/vendors-node_modules_theia_core_shared_phosphor_domutils_index_js-node_modules_theia_monaco-e-1a4eba.js:154858:15)
    at LanguagesMainImpl.$setLanguageConfiguration (http://localhost:3000/vendors-node_modules_theia_plugin-ext_lib_hosted_browser_hosted-plugin_js.js:5295:48)
    at RPCProtocolImpl.doInvokeHandler (http://localhost:3000/vendors-node_modules_theia_plugin-ext_lib_common_disposable-util_js-node_modules_theia_plugin-cba7ee.js:1394:23)
    at RPCProtocolImpl.invokeHandler (http://localhost:3000/vendors-node_modules_theia_plugin-ext_lib_common_disposable-util_js-node_modules_theia_plugin-cba7ee.js:1379:41)
    at RPCProtocolImpl.receiveRequest (http://localhost:3000/vendors-node_modules_theia_plugin-ext_lib_common_disposable-util_js-node_modules_theia_plugin-cba7ee.js:1343:33)
    at RPCProtocolImpl.receiveOneMessage (http://localhost:3000/vendors-node_modules_theia_plugin-ext_lib_common_disposable-util_js-node_modules_theia_plugin-cba7ee.js:1307:26)
    at http://localhost:3000/vendors-node_modules_theia_plugin-ext_lib_common_disposable-util_js-node_modules_theia_plugin-cba7ee.js:1231:48
    at http://localhost:3000/bundle.js:58560:69
    at CallbackList.invoke (http://localhost:3000/bundle.js:58566:26)
    at Emitter.fire (http://localhost:3000/bundle.js:58681:29)
```

Where would a new language be registered? Could it also be an issue with file/folder permissions?

**Reply by fipro78** [reply]
Additional info:

When trying to open a .py file I can now also see this error:

```
2022-08-03T13:26:23.034Z root ERROR Failed to parse data from " /home/node/.theia/plugin-storage/global-state.json ". Reason: SyntaxError: /home/node/.theia/plugin-storage/global-state.json: Unexpected token N in JSON at position 2876
    at JSON.parse (<anonymous>)
    at /workspaces/rds_visualization_app/node_modules/jsonfile/index.js:33:18
    at /workspaces/rds_visualization_app/node_modules/graceful-fs/graceful-fs.js:123:16
    at FSReqCallback.readFileAfterClose [as oncomplete] (internal/fs/read_file_context.js:71:3)
```

Could this be somehow related to https://github.com/eclipse-theia/theia/issues/8384 ? And if so, how can this be fixed? There doesn't seem to be a fix available.

**Reply by msujew** [reply]
> I finally noticed the real issue. It wasn't a proxy issue. I was mislead by this part in the error "JsonRpcProxyFactory.onRequest". The error was that the folder ~/.theia/extensions couldn't be created. After I created it manually from a terminal the installation worked.

Ah, yes we accidentally broke this one, though it is already fixed on [`master`](https://github.com/eclipse-theia/theia/pull/11486).

> Not sure how to debug. I'm still trying to get my head around Theia itself and try to create a usable Theia application. If you have a pointer how to start debugging, please let me know.

The easiest way is probably to follow our [development guide](https://github.com/eclipse-theia/theia/blob/master/doc/Developing.md#debugging), which outlines how one would debug a Theia application.

> Where would a new language be registered? Could it also be an issue with file/folder permissions?

That's weird, you shouldn't have to do that as the extension should register the language itself. Are you building your own Theia app/extension or are you running Theia from source? A repo to replicate this  behavior would be great, as I can't reproduce it using our own repo.

**Reply by fipro78** [reply]
> That's weird, you shouldn't have to do that as the extension should register the language itself. Are you building your own Theia app/extension or are you running Theia from source? A repo to replicate this behavior would be great, as I can't reproduce it using our own repo.

Well, as posted I have a simple setup that creates a Theia application. I am using VS Code with Remote Containers (Docker file posted above) and a Theia app (package.json posted above). But I can try to create a repo with the same setup simplified so it is easier to reproduce the issue.

**Reply by fipro78** [reply]
@msujew 
I have created a new project on a new Windows PC to verify that I am not running into a local setup issue. I use Windows 10 and VS Code with the Remote Container Extension. I created a widget project using the Yeoman Theia Extension Generator.

Once the container is started, trigger the build via `yarn`
Once the build is finished, start the application via `yarn start:browser`

Open a browser on http://localhost:3000
After a folder was opened you should see that:
- the Python Extension is installed as build-in extension
- creating a new .py file will not recognize Python as language
- Setting the Python interpreter will log several errors in the host VS Code terminal

At least this is happening for me on a Windows PC as host. Not sure if this is somehow related.

The repo with the sources to reproduce are located here: https://github.com/fipro78/theia-playground
